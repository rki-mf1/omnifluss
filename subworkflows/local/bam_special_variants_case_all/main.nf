
include { LOFREQ_CALLPARALLEL as LOFREQ_CALLPARALLEL_SPEC_CASE }    from '../../../modules/nf-core/lofreq/callparallel/main'
include { LOFREQ_FILTER }                                           from '../../../modules/nf-core/lofreq/filter/main'
include { INV_RESCUE_VARIANTS_PYTHON }                              from '../../../modules/local/inv_rescue_variants_python/main'


workflow BAM_SPECIAL_VARIANTS_CASE_ALL {

    take:
    ch_iqbam        // channel: [ val(meta), [ bam ]   ]
    ch_iqbam_bai    // channel: [ val(meta), [ bai ]   ]
    ch_ref          // channel: [ val(meta), [ fasta ] ]
    ch_ref_index    // channel: [ val(meta), [ fai ]   ]


    main:
    ch_vcf      = Channel.empty()
    ch_bed      = Channel.empty()
    ch_versions = Channel.empty()

    /****************************************************************/
    /* STEP 0: Format input Channels for LOFREQ_CALLPARALLEL        */
    /****************************************************************/
    ch_iqbam_cpy        = ch_iqbam.map{ meta, iqbam -> return [meta.id, meta, iqbam ] }
    ch_iqbam_bai_cpy    = ch_iqbam_bai.map{ meta, iqbai -> return [meta.id, meta, iqbai ] }
    ch_ref_cpy          = ch_ref.map{ meta, fasta -> return [meta.id, meta, fasta] }
    ch_ref_index_cpy    = ch_ref_index.map{ meta, fai -> return [meta.id, meta, fai] }

    ch_lofreq_callparallel_input = ch_iqbam_cpy.join(ch_iqbam_bai_cpy).join(ch_ref_cpy).join(ch_ref_index_cpy)
        .multiMap{ _sample_id, meta, iqbam, _meta2, iqbai, meta3, fasta, meta4, fai ->
            ch_input_sample: [meta, iqbam, iqbai, [] ]  // needs empty optional parameter here
            ch_ref: [meta3, fasta]
            ch_ref_index: [meta4, fai]
            }

    /*******************************************************************/
    /* STEP 1: LOFREQ_CALLPARALLEL variant calling                     */
    /*******************************************************************/
    LOFREQ_CALLPARALLEL_SPEC_CASE (
        ch_lofreq_callparallel_input.ch_input_sample,
        ch_lofreq_callparallel_input.ch_ref,
        ch_lofreq_callparallel_input.ch_ref_index
    )
    ch_versions = ch_versions.mix(LOFREQ_CALLPARALLEL_SPEC_CASE.out.versions.first())

    /*******************************************************************/
    /* STEP 2: LOFREQ_FILTER filters the variant set; see config       */
    /*******************************************************************/
    LOFREQ_FILTER (
        LOFREQ_CALLPARALLEL_SPEC_CASE.out.vcf
    )
    ch_vcf      = LOFREQ_FILTER.out.vcf
    ch_versions = ch_versions.mix(LOFREQ_FILTER.out.versions.first())

    /***************************************************************************************************************/
    /* STEP 3: INV_RESCUE_VARIANTS_PYTHON generates a bed file with HQ variant sites to be masked in the consensus */
    /***************************************************************************************************************/
    INV_RESCUE_VARIANTS_PYTHON (
        ch_vcf
    )
    ch_bed      = INV_RESCUE_VARIANTS_PYTHON.out.bed
    ch_versions = ch_versions.mix(INV_RESCUE_VARIANTS_PYTHON.out.versions.first())


    emit:
    vcf      = ch_vcf       // channel: [ val(meta), [ vcf ] ]
    bed      = ch_bed       // channel: [ val(meta), [ bed ] ]
    versions = ch_versions  // channel: [ versions.yml ]
}

