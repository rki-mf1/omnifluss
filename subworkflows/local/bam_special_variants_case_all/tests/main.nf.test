
nextflow_workflow {

    name "Test Subworkflow BAM_SPECIAL_VARIANTS_CASE_ALL"
    script "../main.nf"
    config "subworkflows/local/bam_special_variants_case_all/tests/nextflow.config"
    workflow "BAM_SPECIAL_VARIANTS_CASE_ALL"

    tag "subworkflows"
    tag "subworkflows/local"
    tag "subworkflows/local/bam_special_variants_case_all"

    tag "lofreq"
    tag "lofreq/callparallel"
    tag "lofreq/filter"

    tag "inv_rescue_variants_python"


    test("sarscov2 - bam") {

        when {
            workflow {
                """
                input[0] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                            ])
                input[1] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                            ])
                input[2] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                            ])
                input[3] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                            ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert file(workflow.out.vcf.flatten()[1]).name == "test.filtered.vcf.gz" },
                { assert snapshot(workflow.out["bed"]).match() }
            )
        }
    }


    test("sarscov2 - bam - stub ") {

        tag "stub"
        options '-stub'

        when {
            workflow {
                """
                input[0] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam', checkIfExists: true)
                            ])
                input[1] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/bam/test.paired_end.sorted.bam.bai', checkIfExists: true)
                            ])
                input[2] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true)
                            ])
                input[3] =  Channel.value([
                                [ id:'test' ],
                                file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta.fai', checkIfExists: true)
                            ])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success },
                { assert snapshot(workflow.out).match() }
            )
        }
    }
}
