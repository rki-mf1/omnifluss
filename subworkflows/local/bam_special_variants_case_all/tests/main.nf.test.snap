{
    "sarscov2 - bam - stub ": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test.filtered.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test.special_case_variant_mask.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,0ec62ff385fc621e2b8c6797a951c9a9",
                    "versions.yml:md5,1b50ae01e6421949a8afc86a239f836f",
                    "versions.yml:md5,3f03ba9f662c89ac04cb9fefec09edca"
                ],
                "bed": [
                    [
                        {
                            "id": "test"
                        },
                        "test.special_case_variant_mask.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "vcf": [
                    [
                        {
                            "id": "test"
                        },
                        "test.filtered.vcf.gz:md5,68b329da9893e34099c7d8ad5cb9c940"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,0ec62ff385fc621e2b8c6797a951c9a9",
                    "versions.yml:md5,1b50ae01e6421949a8afc86a239f836f",
                    "versions.yml:md5,3f03ba9f662c89ac04cb9fefec09edca"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2025-02-13T16:48:05.41221137"
    },
    "sarscov2 - bam": {
        "content": [
            [
                [
                    {
                        "id": "test"
                    },
                    "test.special_case_variant_mask.bed:md5,d41d8cd98f00b204e9800998ecf8427e"
                ]
            ]
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.04.1"
        },
        "timestamp": "2024-08-01T10:35:02.146362427"
    }
}